File: android/data/repositories/src/main/kotlin/com/gemwallet/android/data/repositories/gemstone/BannerStore.kt
package com.gemwallet.android.data.repositories.gemstone

import com.gemwallet.android.data.service.store.database.BannersDao
import com.gemwallet.android.data.service.store.database.entities.DbBanner
import com.gemwallet.android.serializer.decodeJson
import com.gemwallet.android.serializer.toJson
import com.wallet.core.primitives.BannerEvent
import com.wallet.core.primitives.Chain
import uniffi.gemstone.BannerState
import uniffi.gemstone.GemBannerKey
import uniffi.gemstone.GemBannerStore
import uniffi.gemstone.bannerIdentifier

class GemstoneBannerStore(
    private val bannersDao: BannersDao,
) : GemBannerStore {

    override suspend fun getState(key: GemBannerKey): BannerState? =
        bannersDao.getBanner(bannerIdentifier(key))?.state?.toJson()

    override suspend fun setState(key: GemBannerKey, state: BannerState) {
        val record = key.toRecord(state)
        if (bannersDao.getBanner(record.id)?.state != record.state) {
            bannersDao.saveBanner(record)
        }
    }

    override suspend fun addBanners(keys: List<GemBannerKey>, state: BannerState) {
        bannersDao.addBanners(keys.map { it.toRecord(state) })
    }

    private fun GemBannerKey.toRecord(state: BannerState) = DbBanner(
        id = bannerIdentifier(this),
        walletId = walletId,
        assetId = assetId,
        chain = chain?.let { chain -> Chain.entries.firstOrNull { it.string == chain } },
        event = event.decodeJson<BannerEvent>(),
        state = state.decodeJson(),
    )
}

File: ios/Packages/GemstoneServices/Sources/Stores/BannerStore.swift
// Copyright (c). Gem Wallet. All rights reserved.

import Foundation
import typealias Gemstone.BannerState
import struct Gemstone.GemBannerKey
import protocol Gemstone.GemBannerStore
import GemstonePrimitives
import Primitives
import Store

public final class GemstoneBannerStore: GemBannerStore, @unchecked Sendable {
    private let store: BannerStore

    public init(store: BannerStore) {
        self.store = store
    }

    public func getState(key: GemBannerKey) async throws -> Gemstone.BannerState? {
        try store.getBanner(id: key.identifier())
            .map { $0.state.map() }
    }

    public func setState(key: GemBannerKey, state: Gemstone.BannerState) async throws {
        let state = state.map()
        try store.addBanners([newBanner(key: key, state: state)])
        try store.updateState(key.identifier(), state: state)
    }

    public func addBanners(keys: [GemBannerKey], state: Gemstone.BannerState) async throws {
        let state = state.map()
        try store.addBanners(keys.map { try newBanner(key: $0, state: state) })
    }

    private func newBanner(key: GemBannerKey, state: Primitives.BannerState) throws -> NewBanner {
        try NewBanner(
            id: key.identifier(),
            walletId: key.walletId,
            assetId: key.assetId.map { try Primitives.AssetId(id: $0) },
            event: key.event.map(),
            state: state,
        )
    }
}
